nifi-nar-bundles/nifi-hubspot-bundle/nifi-hubspot-processors/src/main/java/org/apache/nifi/processors/hubspot/GetHubSpot.java

```diff
@@ -187,78 +257,158 @@ private String getResponseBodyAsString(final ProcessContext context, final HttpR
-        return webClientServiceProvider.getWebClientService()
-                .get()
-                .uri(uri)
-                .header("Authorization", "Bearer " + accessToken)
-                .retrieve();
+    private HttpResponseEntity getHttpResponseEntity(final String accessToken, final URI uri, final String filters) {
+        final InputStream inputStream = IOUtils.toInputStream(filters, StandardCharsets.UTF_8);
+        try {
+            return webClientServiceProvider.getWebClientService()
+                    .post()
+                    .uri(uri)
+                    .header("Authorization", "Bearer " + accessToken)
+                    .header("Content-Type", "application/json")
+                    .body(inputStream, OptionalLong.of(inputStream.available()))
+                    .retrieve();
+        } catch (IOException e) {
+            throw new ProcessException("Could not transform incremental filters to input stream", e);
+        }
     }
 
-    private URI createUri(final ProcessContext context, final StateMap state) {
-        final String path = context.getProperty(OBJECT_TYPE).getValue();
-        final HttpUriBuilder uriBuilder = getBaseUri(context);
+    private String createIncrementalFilters(final ProcessContext context, final Map<String, String> stateMap) {
+        final String limit = context.getProperty(RESULT_LIMIT).getValue();
+        final String objectType = context.getProperty(OBJECT_TYPE).getValue();
+        final HubSpotObjectType hubSpotObjectType = OBJECT_TYPE_LOOKUP_MAP.get(objectType);
+        final Long incrDelayMs = context.getProperty(INCREMENTAL_DELAY).asTimePeriod(TimeUnit.MILLISECONDS);
+        final String cursorKey = String.format(CURSOR_KEY_PATTERN, objectType);
 
-        final boolean isLimitSet = context.getProperty(RESULT_LIMIT).evaluateAttributeExpressions().isSet();
-        if (isLimitSet) {
-            final String limit = context.getProperty(RESULT_LIMIT).getValue();
-            uriBuilder.addQueryParameter(LIMIT_PARAMETER, limit);
+        final ObjectNode root = OBJECT_MAPPER.createObjectNode();
+        root.put("limit", limit);
+
+        final String cursor = stateMap.get(cursorKey);
+        if (cursor != null && !NO_PAGING.equals(cursor)) {
+            root.put(PAGING_CURSOR, stateMap.get(cursorKey));
         }
+        final boolean isIncremental = context.getProperty(IS_INCREMENTAL).asBoolean();
+        if (isIncremental) {
+            final String initialStartTimeValue = context.getProperty(INCREMENTAL_INITIAL_START_TIME).getValue();
+            String initialStartTimeEpoch = getInitialStartTimeEpoch(initialStartTimeValue);
+            final String hubspotSpecificIncrementalFieldName = hubSpotObjectType.getLastModifiedDateType().getValue();
+            final String lastStartTime = stateMap.get(START_INCREMENTAL_KEY);
+            final String lastEndTime = stateMap.getOrDefault(END_INCREMENTAL_KEY, initialStartTimeEpoch);
```

Comment:
It is quite hard to understand why `lastEndTime` is initialized with `initialStartTimeEpoch`.
It would be better to assign the End Time value form the state only (which can be null), and use the Initial Start Time when `currentStartTime` is calculated in line 348:
```
currentStartTime = lastEndTime != null ? lastEndTime : getInitialStartTimeEpoch(initialStartTimeValue)`
```
In this case the epoch is calculated only when needed (once).

Reply:
Very good idea, thank you!